src/lib/api/supabase/projects/api.ts

```diff
@@ -0,0 +1,89 @@
+import { createServerSupabaseClient } from '@supabase/auth-helpers-nextjs'
+import { DBProject, DBProjectQueryOpts } from 'models/dbProject'
+import { Json } from 'models/json'
+import { NextApiRequest, NextApiResponse } from 'next'
+import { Database } from 'types/database.types'
+import { dbProjects } from '../clients'
+
+/**
+ * Exhaustively queries all database projects.
+ *
+ * @returns Promise containing all projects from database
+ */
+export function dbpQueryAll() {
+  return dbProjects.select('*').then(
+    res => ({ data: res.data as Json<DBProject>[], error: undefined }),
+    error => ({ data: [] as Json<DBProject>[], error }),
+  )
+}
+
+/**
+ * Writes records to database
+ *
+ * @param records Projects to write
+ */
+export async function writeDBProjects(records: Json<DBProject>[]) {
+  // Sanity check that all IDs are defined
+  const missingIds = records.filter(r => r.id === undefined || r.id === null)
+
+  if (missingIds.length) {
+    throw new Error(
+      `${missingIds.length} records are missing an id: ${missingIds
+        .map(p => JSON.stringify(p))
+        .join(', ')}`,
+    )
+  }
+
+  const _updatedAt = Date.now()
+
+  const queue = records.map(r => ({
+    ...r,
+    _updatedAt,
+
+    // hacky? fix to avoid undefined column vals
```

Comment:
supabase gets mad if we write any projects with undefined values (`"Error: All objects must have matching keys"`). so we do this to set any possible undefined values to `null`